<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Metabolic Labeling for Forecasts · RNAForecaster</title><script data-outdated-warner src="../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.045/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.24/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../">RNAForecaster</a></span></div><form class="docs-search" action="../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../">Home</a></li><li><a class="tocitem" href="../training/">Training and Forecasting</a></li><li><a class="tocitem" href="../splicedData/">Predictions from Splicing Data</a></li><li class="is-active"><a class="tocitem" href>Metabolic Labeling for Forecasts</a></li><li><a class="tocitem" href="../functions/">Functions</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Metabolic Labeling for Forecasts</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Metabolic Labeling for Forecasts</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/rossinerbe/RNAForecaster.jl/blob/main/docs/src/forecastMLData.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Metabolic-Labeling-for-Forecasts"><a class="docs-heading-anchor" href="#Metabolic-Labeling-for-Forecasts">Metabolic Labeling for Forecasts</a><a id="Metabolic-Labeling-for-Forecasts-1"></a><a class="docs-heading-anchor-permalink" href="#Metabolic-Labeling-for-Forecasts" title="Permalink"></a></h1><p>One way to get the two time points needed to train RNAForecaster is to use metabolic labeling transcriptomic profiling techniques such as scEU-seq. In these techniques, cells are incubated with modified uridine for a set period of time, after which they are harvested for scRNA sequencing. The transcripts with incorporated uridine are labeled, and we therefore know they were transcribed in the labeling time period. We also know the other transcripts were transcribed earlier.</p><p>We can use these labeled and unlabeled transcripts to generate our two input matrices. To get a clean time point t=0 matrix of total expression that we can use with the total expression at time point t=1 (when the cells were harvested) we do need to estimate the number of degraded transcripts in the labeling period, which we calculate via the slope between the labeled and total transcripts (see <a href="https://doi.org/10.1016/j.cell.2021.12.045">Qiu et al. 2022</a>).</p><p>Here we use a subset of a scEUseq dataset from hTERT immortalized retinal pigment epithelium cells, published by <a href="https://doi.org/10.1126/science.aax3072">Battich et al. 2020</a>.</p><pre><code class="language-julia hljs">using JLD2
rpeExampleData = load_object(&quot;data/rpeExampleData.jld2&quot;);
labeledData = Float32.(rpeExampleData[1])
unlabeledData = Float32.(rpeExampleData[2])
totalData = Float32.(rpeExampleData[3])
geneNames = string.(rpeExampleData[4])
labelingTime = vec(rpeExampleData[5])

t0Estimate = estimateT0LabelingData(labeledData, totalData, unlabeledData, labelingTime)</code></pre><p>For this tutorial, we will subset to cells labeled for 60 minutes, enabling us to forecast in one hour increments.</p><pre><code class="language-julia hljs">hourCells = findall(x-&gt;x == 60.0, labelingTime)
t0_60min = t0Estimate[:,hourCells]
t1_60min = totalData[:,hourCells]

#log normalize
t0_60min = log1p.(t0_60min)
t1_60min = log1p.(t1_60min)

#train RNAForecaster
trainedNetwork = trainRNAForecaster(t0_60min, t1_60min, trainingProp = 1.0,
 learningRate = 0.0001, batchsize = 25);</code></pre><p>If we wanted to train the network using the GPU, we would run <code>trainedNetwork = trainRNAForecaster(t0_60min, t1_60min, trainingProp = 1.0,  learningRate = 0.0001, batchsize = 25, useGPU = true)</code></p><p>Note on GPU errors: on some systems an error can arise related to the way julia and CUDA handle memory. If you encounter unexpected Out of GPU Memory errors, try setting the environment variable <code>JULIA_CUDA_MEMORY_POOL = none</code></p><p>Now, we can forecast expression predictions 24 hours into the future, and get predictions at each hour.</p><pre><code class="language-julia hljs">futureExpressionPreds = predictCellFutures(trainedNetwork[1], t1_60min, 24)</code></pre><p>Predictions can also be performed on the GPU <code>futureExpressionPreds = predictCellFutures(trainedNetwork[1], t1_60min, 24,    useGPU = true)</code></p><p>The predictions can also be conditioned on arbitrary perturbations in gene expression.</p><pre><code class="language-julia hljs">futureExpressionPredsP = predictCellFutures(trainedNetwork[1], t1_60min, 24,
 perturbGenes = geneNames[1:2], geneNames = geneNames,
 perturbationLevels = [2.0f0, 0.0f0])</code></pre><p>Once we have forecast expression levels for each gene, we may want to know which genes expression levels change the most over time, as these are likely to be important in ongoing biological process we are attempting to model. To assay this we simply run <code>mostTimeVariableGenes</code> which outputs a table of genes ordered by the most variable over predicted time points.</p><pre><code class="language-julia hljs">geneOutputTable = mostTimeVariableGenes(futureExpressionPreds, geneNames)</code></pre><p>We can save an RNAForecaster network for later use with:</p><pre><code class="language-julia hljs">saveForecaster(trainedNetwork, &quot;exampleForecaster.jld2&quot;)</code></pre><p>which we can load back into memory using</p><pre><code class="language-julia hljs">loadedNetwork = loadForecaster(&quot;exampleForecaster.jld2&quot;)</code></pre><p>WARNING: If you update Flux.jl or DiffEqFlux.jl, saved networks are not guaranteed to be forward compatible.</p></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../splicedData/">« Predictions from Splicing Data</a><a class="docs-footer-nextpage" href="../functions/">Functions »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.23 on <span class="colophon-date" title="Tuesday 1 November 2022 17:13">Tuesday 1 November 2022</span>. Using Julia version 1.8.2.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
