<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Functions · RNAForecaster</title><script data-outdated-warner src="../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.045/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.24/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../">RNAForecaster</a></span></div><form class="docs-search" action="../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../">Home</a></li><li><a class="tocitem" href="../training/">Training and Forecasting</a></li><li><a class="tocitem" href="../splicedData/">Predictions from Splicing Data</a></li><li><a class="tocitem" href="../forecastMLData/">Metabolic Labeling for Forecasts</a></li><li class="is-active"><a class="tocitem" href>Functions</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Functions</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Functions</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/rossinerbe/RNAForecaster.jl/blob/main/docs/src/functions.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><article class="docstring"><header><a class="docstring-binding" id="RNAForecaster.createEnsembleForecaster-Tuple{Matrix{Float32}, Matrix{Float32}}" href="#RNAForecaster.createEnsembleForecaster-Tuple{Matrix{Float32}, Matrix{Float32}}"><code>RNAForecaster.createEnsembleForecaster</code></a> — <span class="docstring-category">Method</span></header><section><div><p><code>createEnsembleForecaster(expressionDataT0::Matrix{Float32}, expressionDataT1::Matrix{Float32};      nNetworks::Int = 5, trainingProp::Float64 = 0.8,      hiddenLayerNodes::Int = 2*size(expressionDataT0)[1],      shuffleData::Bool = true, seed::Int = 123, learningRate::Float64 = 0.005,      nEpochs::Int = 10, batchsize::Int = 100, checkStability::Bool = false,      iterToCheck::Int = 50, stabilityThreshold::Float32 = 2*maximum(expressionDataT0),      stabilityChecksBeforeFail::Int = 5, useGPU::Bool = false)</code></p><p>Function to train multiple neural ODEs to predict expression, allowing an ensembling of their predicitons, which tends to yield more accurate results on future predictions. This is because stochastic gradient descent yields slightly different solutions when given different random seeds. In the training data these solutions yield almost identical results, but when generalizing to future predictions, the results can diverge substantially. To account for this, we can average across multiple forecasters.</p><p>It is recommended to run this function on a GPU (useGPU = true) or if a GPU is not available run in parallel. To train the neural networks on separate processes call <code>using Distributed addprocs(desiredNumberOfParallelProcesses) @everywhere using RNAForecaster</code></p><p><strong>Required Arguments</strong></p><ul><li>expressionDataT0 - Float32 Matrix of log-normalized expression counts in the format of genes x cells</li><li>expressionDataT1 - Float32 Matrix of log-normalized expression counts in the format</li></ul><p>of genes x cells from a time after expressionDataT0</p><p><strong>Keyword Arguments</strong></p><ul><li>nNetworks - number of networks to train</li><li>trainingProp - proportion of the data to use for training the model, the rest will be</li></ul><p>used for a validation set. If you don&#39;t want a validation set, this value can be set to 1.0</p><ul><li>hiddenLayerNodes - number of nodes in the hidden layer of the neural network</li><li>shuffleData - should the cells be randomly shuffled before training</li><li>seed - random seed</li><li>learningRate - learning rate for the neural network during training</li><li>nEpochs - how many times should the neural network be trained on the data.</li></ul><p>Generally yields small gains in performance, can be lowered to speed up the training process</p><ul><li>batchsize - batch size for training</li><li>checkStability - should the stability of the networks future time predictions be checked,</li></ul><p>retraining the network if unstable?</p><ul><li>iterToCheck - when checking stability, how many future time steps should be predicted?</li><li>stabilityThreshold - when checking stability, what is the maximum gene variance allowable across predictions?</li><li>stabilityChecksBeforeFail - when checking stability, how many times should the network</li></ul><p>be allowed to retrain before an error is thrown? Used to prevent an infinite loop.</p><ul><li>useGPU - use a GPU to train the neural network? highly recommended for large data sets, if available</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/rossinerbe/RNAForecaster.jl/blob/0151583ae30942235017fc2fb120fe8914a03ebb/src/trainRNAForecaster.jl#L280-L325">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="RNAForecaster.estimateT0LabelingData-Tuple{Matrix{Float32}, Matrix{Float32}, Matrix{Float32}, AbstractVector}" href="#RNAForecaster.estimateT0LabelingData-Tuple{Matrix{Float32}, Matrix{Float32}, Matrix{Float32}, AbstractVector}"><code>RNAForecaster.estimateT0LabelingData</code></a> — <span class="docstring-category">Method</span></header><section><div><p>&quot; <code>estimateT0LabelingData(labeledData::Matrix{Float32}, totalData::Matrix{Float32},     unlabeledData::Matrix{Float32}, labelingTime::AbstractVector)</code></p><p>Function to predict total expression level before labeling based on degradation rate estimates. Outputs the estimated time 1 counts matrix.</p><p><strong>Required Arguments</strong></p><ul><li>labeledData - Float32 counts matrix of the labeled counts</li><li>totalData - Float32 counts matrix of combined counts</li><li>unlabeledData - Float32 counts matrix of unlabeled counts</li><li>labelingTime - Vector with the amount of time each cell was labeled for</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/rossinerbe/RNAForecaster.jl/blob/0151583ae30942235017fc2fb120fe8914a03ebb/src/estimateT0.jl#L23-L37">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="RNAForecaster.filterByGeneVar-Tuple{Matrix{Float32}, Matrix{Float32}, Int64}" href="#RNAForecaster.filterByGeneVar-Tuple{Matrix{Float32}, Matrix{Float32}, Int64}"><code>RNAForecaster.filterByGeneVar</code></a> — <span class="docstring-category">Method</span></header><section><div><p><code>filterByGeneVar(t0Counts::Matrix{Float32}, t1Counts::Matrix{Float32},     topGenes::Int)</code></p><p>Filter by gene variance, measured across both count matrices.</p><p><strong>Required Arguments</strong></p><ul><li>t0Counts - Counts matrix for time 1. Should be genes x cells and Float32.</li><li>t1Counts - Counts matrix for time 2. Should be genes x cells and Float32.</li><li>topGenes - the number of top most variable genes to use</li></ul><p><strong>Examples</strong></p><p><code>mat0 = Float32.(randn(50,50)) mat1 = Float32.(randn(50,50)) hvgFilteredCounts = filterByGeneVar(mat0, mat1, 20)</code></p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/rossinerbe/RNAForecaster.jl/blob/0151583ae30942235017fc2fb120fe8914a03ebb/src/utils.jl#L53-L68">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="RNAForecaster.filterByZeroProp-Tuple{Matrix{Float32}, Matrix{Float32}}" href="#RNAForecaster.filterByZeroProp-Tuple{Matrix{Float32}, Matrix{Float32}}"><code>RNAForecaster.filterByZeroProp</code></a> — <span class="docstring-category">Method</span></header><section><div><p><code>filterByZeroProp(t0Counts::Matrix{Float32}, t1Counts::Matrix{Float32},     zeroProp::Float32)</code></p><p>Filter by zero proportion for both genes and cells. Very high sparsity prevents the neural network from achieving a stable solution.</p><p><strong>Required Arguments</strong></p><ul><li>t0Counts - Counts matrix for time 1. Should be genes x cells and Float32.</li><li>t1Counts - Counts matrix for time 2. Should be genes x cells and Float32.</li></ul><p><strong>Keyword Arguments</strong></p><ul><li>zeroProp - proportion of zeroes allowed for a gene or a cell. 0.98f0 by default</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/rossinerbe/RNAForecaster.jl/blob/0151583ae30942235017fc2fb120fe8914a03ebb/src/utils.jl#L4-L17">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="RNAForecaster.findSigRegulation-Tuple{Any, Vector{String}}" href="#RNAForecaster.findSigRegulation-Tuple{Any, Vector{String}}"><code>RNAForecaster.findSigRegulation</code></a> — <span class="docstring-category">Method</span></header><section><div><p><code>findSigRegulation(perturbData, geneNames::Vector{String};     pvalCut::Float64 = 0.05)</code></p><p>Function to get a sorted data frame of whether there exists a statistically significant difference between predicted gene expression before and after perturbation.</p><p><strong>Required Arguments</strong></p><ul><li>perturbData - results from perturbEffectPredictions function</li><li>geneNames - vector of gene names in the order of the input expression data</li><li>pvalCut - p value threshold. Default is 0.05</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/rossinerbe/RNAForecaster.jl/blob/0151583ae30942235017fc2fb120fe8914a03ebb/src/splicedDataPerturbationEffectPredictions.jl#L232-L242">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="RNAForecaster.genePerturbExpressionChanges-Tuple{Any, Vector{String}, String}" href="#RNAForecaster.genePerturbExpressionChanges-Tuple{Any, Vector{String}, String}"><code>RNAForecaster.genePerturbExpressionChanges</code></a> — <span class="docstring-category">Method</span></header><section><div><p><code>genePerturbExpressionChanges(perturbData, geneNames::Vector{String}, perturbGene::String;     genesperturbd::Vector{String} = geneNames)</code></p><p>Function to get a sorted data frame of the predicted effect of a gene perturb on all other genes.</p><p><strong>Required Arguments</strong></p><ul><li>perturbData - results from perturbEffectPredictions function</li><li>geneNames - vector of gene names in the order of the input expression data</li><li>perturbGene - a gene name to query the predicted perturb effect on expression</li></ul><p><strong>Optional Arguments</strong></p><ul><li>genesPerturbed - If less than all the gene perturbs were performed, the ordered names of the perturb genes must be supplied</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/rossinerbe/RNAForecaster.jl/blob/0151583ae30942235017fc2fb120fe8914a03ebb/src/splicedDataPerturbationEffectPredictions.jl#L149-L162">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="RNAForecaster.geneResponseToPerturb-Tuple{Any, Vector{String}, String}" href="#RNAForecaster.geneResponseToPerturb-Tuple{Any, Vector{String}, String}"><code>RNAForecaster.geneResponseToPerturb</code></a> — <span class="docstring-category">Method</span></header><section><div><p><code>geneResponseToPerturb(perturbData, geneNames::Vector{String}, geneOfInterest::String;     genesPerturbed::Vector{String} = geneNames)</code></p><p>Function to get a sorted data frame of the predicted effect of all other gene perturbations on a particular gene of interest.</p><p><strong>Required Arguments</strong></p><ul><li>perturbData - results from perturbEffectPredictions function</li><li>geneNames - vector of gene names in the order of the input expression data</li><li>geneOfInterest - a gene name to query</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/rossinerbe/RNAForecaster.jl/blob/0151583ae30942235017fc2fb120fe8914a03ebb/src/splicedDataPerturbationEffectPredictions.jl#L195-L205">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="RNAForecaster.loadForecaster-Tuple{String}" href="#RNAForecaster.loadForecaster-Tuple{String}"><code>RNAForecaster.loadForecaster</code></a> — <span class="docstring-category">Method</span></header><section><div><p><code>loadForecaster(fileName::String, inputNodes::Int, hiddenLayerNodes::Int)</code></p><p>Recreates a previously saved neural network.</p><p>Note: if for some reason you are loading the network and have not first loaded the DiffEqFlux and DifferentialEquations packages (normally should be loaded when loading RNAForecaster.jl) then the network will not work, even if you load the required packages afterwards.</p><p><strong>Required Arguments</strong></p><ul><li>fileName - file name where the parameters are saved</li><li>inputNodes - number of input nodes in the network. Should be the same as the number</li></ul><p>of genes in the data the network was trained on</p><ul><li>hiddenLayerNodes - number of hidden layer nodes in the network</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/rossinerbe/RNAForecaster.jl/blob/0151583ae30942235017fc2fb120fe8914a03ebb/src/utils.jl#L95-L110">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="RNAForecaster.mostTimeVariableGenes-Tuple{AbstractArray{Float32}, Vector{String}}" href="#RNAForecaster.mostTimeVariableGenes-Tuple{AbstractArray{Float32}, Vector{String}}"><code>RNAForecaster.mostTimeVariableGenes</code></a> — <span class="docstring-category">Method</span></header><section><div><p><code>mostTimeVariableGenes(cellFutures::AbstractArray{Float32}, geneNames::Vector{String};      statType = &quot;mean&quot;)</code></p><p>For each cell, takes the predicted expression levels of each gene over time and finds the variance with respect to predicted time points. Then get the mean/median for each gene&#39;s variance across cells for each gene.</p><p>Outputs a sorted DataFrame containing gene names and the variances over predicted time.</p><p><strong>Required Arguments</strong></p><ul><li>cellFutures - a 3D tensor of gene expression over time; the output from predictCellFutures</li><li>geneNames - a vector of gene names corresponding to the order of the genes in cellFutures</li></ul><p><strong>Optional Arguments</strong></p><ul><li>statType - How to summarize the gene variances. Valid options are &quot;mean&quot; or &quot;median&quot;</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/rossinerbe/RNAForecaster.jl/blob/0151583ae30942235017fc2fb120fe8914a03ebb/src/makeRecursivePredictions.jl#L171-L186">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="RNAForecaster.perturbEffectPredictions-Tuple{Any, Matrix{Float32}, Int64}" href="#RNAForecaster.perturbEffectPredictions-Tuple{Any, Matrix{Float32}, Int64}"><code>RNAForecaster.perturbEffectPredictions</code></a> — <span class="docstring-category">Method</span></header><section><div><p><code>perturbEffectPredictions(trainedNetwork, splicedData::Matrix{Float32}, nCells::Int;      perturbGenes::Vector{String} = Vector{String}(undef, 0), geneNames::Vector{String} = Vector{String}(undef,0),      seed::Int=123)</code></p><p>Based on spliced/unspliced counts, predict the immediate transcriptomic effect of any or all single gene perturbations. Outputs a tuple containing the cells used for prediction, the expression predictions, the gene wise differences, and the cell-wise euclidean distances for each perturbation.</p><p>This function is capable of running on multiple parallel processes using Distributed.jl. Call addprocs(n) before running the function to add parallel workers, where n is the number of additional processes desired.</p><p><strong>Required Arguments</strong></p><ul><li>trainedNetwork - trained neuralODE from trainRNAForecaster</li><li>splicedData - log normalized spliced counts matrix. Must be in Float32 format</li><li>nCells - how many cells from the data should be used for prediction of perturb effect.</li></ul><p>Higher values will increase computational time required.</p><p><strong>Optional Arguments</strong></p><ul><li>perturbGenes - list of genes to simulate a perturbation of. By default all genes are used</li><li>geneNames - if providing a subset of the genes to perturb, a vector of gene names to</li></ul><p>match against, in the order of splicedData</p><ul><li>perturbLevels - list of perturbation levels to use for each perturbed gene. By default</li></ul><p>all genes are set to zero, simulating a KO.</p><ul><li>seed - Random seed for reproducibility on the cells chosen for prediction</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/rossinerbe/RNAForecaster.jl/blob/0151583ae30942235017fc2fb120fe8914a03ebb/src/splicedDataPerturbationEffectPredictions.jl#L1-L27">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="RNAForecaster.predictCellFutures-Tuple{Any, Matrix{Float32}, Int64}" href="#RNAForecaster.predictCellFutures-Tuple{Any, Matrix{Float32}, Int64}"><code>RNAForecaster.predictCellFutures</code></a> — <span class="docstring-category">Method</span></header><section><div><p><code>predictCellFutures(trainedNetwork, expressionData::Matrix{Float32}, tSteps::Int;      perturbGenes::Vector{String} = Vector{String}(undef,0), geneNames::Vector{String} = Vector{String}(undef,0),      perturbationLevels::Vector{Float32} = Vector{Float32}(undef,0),      enforceMaxPred::Bool = true, maxPrediction::Float32 = 2*maximum(expressionData))</code></p><p>Function to make future expression predictions using a trained neural ODE outputs a 3d tensor containing a predicted expression counts matrix for the cell at each time step</p><p><strong>Required Arguments</strong></p><ul><li>trainedNetwork - the trained neural ODE, from the trainRNAForecaster function</li><li>expressionData - the initial expression states that should be used to make predictions from</li><li>tSteps - how many future time steps should be predicted. (Error will propagate</li></ul><p>with each prediction so predictions will eventually become highly innaccurate at high numbers of time steps)</p><p><strong>Keyword Arguments</strong></p><ul><li>perturbGenes - a vector of gene names that will have their values set to a constant &#39;perturbed&#39; level.</li><li>geneNames - a vector of gene names in the order of the rows of the expressionData.</li></ul><p>Used only when simulating perturbations.</p><ul><li>perturbationLevels - a vector of Float32, corresponding to the level each perturbed</li></ul><p>gene&#39;s expression should be set at.</p><ul><li>enforceMaxPred - should a maximum allowed prediction be enforced? This is used</li></ul><p>to represent prior knowledge about what sort of expression values are remotely reasonable predictions.</p><ul><li>maxPrediction - if enforcing a maximum prediction, what should the value be?</li></ul><p>2 times the maximum of the input expression data by default (in log space).</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/rossinerbe/RNAForecaster.jl/blob/0151583ae30942235017fc2fb120fe8914a03ebb/src/makeRecursivePredictions.jl#L1-L25">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="RNAForecaster.saveForecaster-Tuple{Any, String}" href="#RNAForecaster.saveForecaster-Tuple{Any, String}"><code>RNAForecaster.saveForecaster</code></a> — <span class="docstring-category">Method</span></header><section><div><p><code>saveForecaster(trainedModel, fileName::String)</code></p><p>Saves the parameters from the neural network after training.</p><p><strong>Required Arguments</strong></p><ul><li>trainedModel - the trained neural network from trainRNAForecaster - just the network</li></ul><p>don&#39;t include the loss results</p><ul><li>fileName - fileName to save the parameters to. Must end in .jld2</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/rossinerbe/RNAForecaster.jl/blob/0151583ae30942235017fc2fb120fe8914a03ebb/src/utils.jl#L81-L90">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="RNAForecaster.totalPerturbImpact-Tuple{Any, Vector{String}}" href="#RNAForecaster.totalPerturbImpact-Tuple{Any, Vector{String}}"><code>RNAForecaster.totalPerturbImpact</code></a> — <span class="docstring-category">Method</span></header><section><div><p><code>totalPerturbImpact(perturbData, geneNames::Vector{String})</code></p><p>Function to yield a sorted data frame of the size of the predicted effect of a perturbation on the cellular transcriptome. Intended to serve as a measure of more or less impactful gene perturbations.</p><p><strong>Required Arguments</strong></p><ul><li>perturbData - results from perturbEffectPredictions function</li><li>geneNames - vector of gene names in the order of the input expression data.</li></ul><p>Should only include perturbed genes</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/rossinerbe/RNAForecaster.jl/blob/0151583ae30942235017fc2fb120fe8914a03ebb/src/splicedDataPerturbationEffectPredictions.jl#L124-L135">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="RNAForecaster.trainRNAForecaster-Tuple{Matrix{Float32}, Matrix{Float32}}" href="#RNAForecaster.trainRNAForecaster-Tuple{Matrix{Float32}, Matrix{Float32}}"><code>RNAForecaster.trainRNAForecaster</code></a> — <span class="docstring-category">Method</span></header><section><div><p><code>trainRNAForecaster(expressionDataT0::Matrix{Float32}, expressionDataT1::Matrix{Float32};      trainingProp::Float64 = 0.8, hiddenLayerNodes::Int = 2*size(expressionDataT0)[1],      shuffleData::Bool = true, seed::Int = 123, learningRate::Float64 = 0.005,      nEpochs::Int = 10, batchsize::Int = 100, checkStability::Bool = false, iterToCheck::Int = 50,      stabilityThreshold::Float32 = 2*maximum(expressionDataT0), stabilityChecksBeforeFail::Int = 5,      useGPU::Bool = false)</code></p><p>Function to train RNAForecaster based on expression data. Main input is two matrices representing expression data from two different time points in the same cell. This can be either based on splicing or metabolic labeling currently. Each should be log normalized and have genes as rows and cells as columns.</p><p><strong>Required Arguments</strong></p><ul><li>expressionDataT0 - Float32 Matrix of log-normalized expression counts in the format of genes x cells</li><li>expressionDataT1 - Float32 Matrix of log-normalized expression counts in the format</li></ul><p>of genes x cells from a time after expressionDataT0</p><p><strong>Keyword Arguments</strong></p><ul><li>trainingProp - proportion of the data to use for training the model, the rest will be</li></ul><p>used for a validation set. If you don&#39;t want a validation set, this value can be set to 1.0</p><ul><li>hiddenLayerNodes - number of nodes in the hidden layer of the neural network</li><li>shuffleData - should the cells be randomly shuffled before training</li><li>seed - random seed</li><li>learningRate - learning rate for the neural network during training</li><li>nEpochs - how many times should the neural network be trained on the data.</li></ul><p>Generally yields small gains in performance, can be lowered to speed up the training process</p><ul><li>batchsize - batch size for training</li><li>checkStability - should the stability of the networks future time predictions be checked,</li></ul><p>retraining the network if unstable?</p><ul><li>iterToCheck - when checking stability, how many future time steps should be predicted?</li><li>stabilityThreshold - when checking stability, what is the maximum gene variance allowable across predictions?</li><li>stabilityChecksBeforeFail - when checking stability, how many times should the network</li></ul><p>be allowed to retrain before an error is thrown? Used to prevent an infinite loop.</p><ul><li>useGPU - use a GPU to train the neural network? highly recommended for large data sets, if available</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/rossinerbe/RNAForecaster.jl/blob/0151583ae30942235017fc2fb120fe8914a03ebb/src/trainRNAForecaster.jl#L135-L169">source</a></section></article></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../forecastMLData/">« Metabolic Labeling for Forecasts</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.23 on <span class="colophon-date" title="Tuesday 1 November 2022 17:13">Tuesday 1 November 2022</span>. Using Julia version 1.8.2.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
