<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Training and Forecasting · RNAForecaster</title><script data-outdated-warner src="../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.045/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.24/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../">RNAForecaster</a></span></div><form class="docs-search" action="../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../">Home</a></li><li class="is-active"><a class="tocitem" href>Training and Forecasting</a><ul class="internal"><li class="toplevel"><a class="tocitem" href="#Training"><span>Training</span></a></li></ul></li><li><a class="tocitem" href="../splicedData/">Predictions from Splicing Data</a></li><li><a class="tocitem" href="../forecastMLData/">Metabolic Labeling for Forecasts</a></li><li><a class="tocitem" href="../functions/">Functions</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Training and Forecasting</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Training and Forecasting</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/rossinerbe/RNAForecaster.jl/blob/main/docs/src/training.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Training-and-Forecasting"><a class="docs-heading-anchor" href="#Training-and-Forecasting">Training and Forecasting</a><a id="Training-and-Forecasting-1"></a><a class="docs-heading-anchor-permalink" href="#Training-and-Forecasting" title="Permalink"></a></h1><p>Training RNAForecaster requires two expression count matrices. These count matrices should be formatted with genes  as rows and cells as columns. Each matrix should represent two different time points from the same cells. This can be accomplished from transcriptomic profiling using spliced and unspliced counts or by using labeled and unlabeled counts from metabolic labeling scRNA-seq protocols such as scEU-seq (see <a href="https://doi.org/10.1126/science.aax3072">Battich et al. 2020</a>).</p><p>Here, we will generate two random matrices for illustrative purposes.</p><pre><code class="language-julia hljs">testT0 = log1p.(Float32.(abs.(randn(10,1000))))
testT1 = log1p.(0.5f0 .* testT0)</code></pre><p>Note that the input matrices are expected to be of type Float32 and log transformed, as shown above.</p><h1 id="Training"><a class="docs-heading-anchor" href="#Training">Training</a><a id="Training-1"></a><a class="docs-heading-anchor-permalink" href="#Training" title="Permalink"></a></h1><p>To train the neural ODE network we call the <code>trainRNAForecaster</code> function.</p><pre><code class="language-julia hljs">testForecaster = trainRNAForecaster(testT0, testT1);</code></pre><p>In the simplest case, we only need to input the matrices, but there are several options provided to modify the training of the neural network, as shown below.</p><article class="docstring"><header><a class="docstring-binding" id="RNAForecaster.trainRNAForecaster" href="#RNAForecaster.trainRNAForecaster"><code>RNAForecaster.trainRNAForecaster</code></a> — <span class="docstring-category">Function</span></header><section><div><p><code>trainRNAForecaster(expressionDataT0::Matrix{Float32}, expressionDataT1::Matrix{Float32};      trainingProp::Float64 = 0.8, hiddenLayerNodes::Int = 2*size(expressionDataT0)[1],      shuffleData::Bool = true, seed::Int = 123, learningRate::Float64 = 0.005,      nEpochs::Int = 10, batchsize::Int = 100, checkStability::Bool = false, iterToCheck::Int = 50,      stabilityThreshold::Float32 = 2*maximum(expressionDataT0), stabilityChecksBeforeFail::Int = 5,      useGPU::Bool = false)</code></p><p>Function to train RNAForecaster based on expression data. Main input is two matrices representing expression data from two different time points in the same cell. This can be either based on splicing or metabolic labeling currently. Each should be log normalized and have genes as rows and cells as columns.</p><p><strong>Required Arguments</strong></p><ul><li>expressionDataT0 - Float32 Matrix of log-normalized expression counts in the format of genes x cells</li><li>expressionDataT1 - Float32 Matrix of log-normalized expression counts in the format</li></ul><p>of genes x cells from a time after expressionDataT0</p><p><strong>Keyword Arguments</strong></p><ul><li>trainingProp - proportion of the data to use for training the model, the rest will be</li></ul><p>used for a validation set. If you don&#39;t want a validation set, this value can be set to 1.0</p><ul><li>hiddenLayerNodes - number of nodes in the hidden layer of the neural network</li><li>shuffleData - should the cells be randomly shuffled before training</li><li>seed - random seed</li><li>learningRate - learning rate for the neural network during training</li><li>nEpochs - how many times should the neural network be trained on the data.</li></ul><p>Generally yields small gains in performance, can be lowered to speed up the training process</p><ul><li>batchsize - batch size for training</li><li>checkStability - should the stability of the networks future time predictions be checked,</li></ul><p>retraining the network if unstable?</p><ul><li>iterToCheck - when checking stability, how many future time steps should be predicted?</li><li>stabilityThreshold - when checking stability, what is the maximum gene variance allowable across predictions?</li><li>stabilityChecksBeforeFail - when checking stability, how many times should the network</li></ul><p>be allowed to retrain before an error is thrown? Used to prevent an infinite loop.</p><ul><li>useGPU - use a GPU to train the neural network? highly recommended for large data sets, if available</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/rossinerbe/RNAForecaster.jl/blob/0151583ae30942235017fc2fb120fe8914a03ebb/src/trainRNAForecaster.jl#L135-L169">source</a></section></article><p>For example, by default RNAForecaster partitions the input data into a training and a validation set. If we want the neural network to be trained on the entire data set, we can set <code>trainingProp = 1.0</code>.</p><p>When using larger data sets, such as a matrix from a normal scRNAseq experiment which may contain thousands of variable genes and tens of thousands of cells, it becomes inefficient to train the network on a CPU. If a GPU is available, setting <code>useGPU = true</code> can massively speed up the training process.</p><p>#Forecasting</p><p>Once we have trained the neural network, we can use it to forecast future expression states. For example, to predict the next fifty time points from our test data, we could run:</p><pre><code class="language-julia hljs">testOut1 = predictCellFutures(testForecaster[1], testT0, 50);</code></pre><p>The predictions can also be conditioned on arbitrary perturbations in gene expression.</p><pre><code class="language-julia hljs">geneNames = [&quot;A&quot;, &quot;B&quot;, &quot;C&quot;, &quot;D&quot;, &quot;E&quot;, &quot;F&quot;, &quot;G&quot;, &quot;H&quot;, &quot;I&quot;, &quot;J&quot;]
testOut2 = predictCellFutures(testForecaster[1], testT0, 50, perturbGenes = [&quot;A&quot;, &quot;B&quot;, &quot;F&quot;],
geneNames = geneNames, perturbationLevels = [1.0f0, 2.0f0, 0.0f0]);</code></pre><p>All options for <code>predictCellFutures</code> are shown here:</p><article class="docstring"><header><a class="docstring-binding" id="RNAForecaster.predictCellFutures" href="#RNAForecaster.predictCellFutures"><code>RNAForecaster.predictCellFutures</code></a> — <span class="docstring-category">Function</span></header><section><div><p><code>predictCellFutures(trainedNetwork, expressionData::Matrix{Float32}, tSteps::Int;      perturbGenes::Vector{String} = Vector{String}(undef,0), geneNames::Vector{String} = Vector{String}(undef,0),      perturbationLevels::Vector{Float32} = Vector{Float32}(undef,0),      enforceMaxPred::Bool = true, maxPrediction::Float32 = 2*maximum(expressionData))</code></p><p>Function to make future expression predictions using a trained neural ODE outputs a 3d tensor containing a predicted expression counts matrix for the cell at each time step</p><p><strong>Required Arguments</strong></p><ul><li>trainedNetwork - the trained neural ODE, from the trainRNAForecaster function</li><li>expressionData - the initial expression states that should be used to make predictions from</li><li>tSteps - how many future time steps should be predicted. (Error will propagate</li></ul><p>with each prediction so predictions will eventually become highly innaccurate at high numbers of time steps)</p><p><strong>Keyword Arguments</strong></p><ul><li>perturbGenes - a vector of gene names that will have their values set to a constant &#39;perturbed&#39; level.</li><li>geneNames - a vector of gene names in the order of the rows of the expressionData.</li></ul><p>Used only when simulating perturbations.</p><ul><li>perturbationLevels - a vector of Float32, corresponding to the level each perturbed</li></ul><p>gene&#39;s expression should be set at.</p><ul><li>enforceMaxPred - should a maximum allowed prediction be enforced? This is used</li></ul><p>to represent prior knowledge about what sort of expression values are remotely reasonable predictions.</p><ul><li>maxPrediction - if enforcing a maximum prediction, what should the value be?</li></ul><p>2 times the maximum of the input expression data by default (in log space).</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/rossinerbe/RNAForecaster.jl/blob/0151583ae30942235017fc2fb120fe8914a03ebb/src/makeRecursivePredictions.jl#L1-L25">source</a></section></article><p>Once we have forecast expression levels for each gene, we may want to know which genes expression levels change the most over time, as these are likely to be important in ongoing biological process we are attempting to model. To assay this we simply run <code>mostTimeVariableGenes</code> which outputs a table of genes ordered by the most variable over predicted time points.</p><pre><code class="language-julia hljs">geneOutputTable = mostTimeVariableGenes(testOut1, geneNames)</code></pre><article class="docstring"><header><a class="docstring-binding" id="RNAForecaster.mostTimeVariableGenes" href="#RNAForecaster.mostTimeVariableGenes"><code>RNAForecaster.mostTimeVariableGenes</code></a> — <span class="docstring-category">Function</span></header><section><div><p><code>mostTimeVariableGenes(cellFutures::AbstractArray{Float32}, geneNames::Vector{String};      statType = &quot;mean&quot;)</code></p><p>For each cell, takes the predicted expression levels of each gene over time and finds the variance with respect to predicted time points. Then get the mean/median for each gene&#39;s variance across cells for each gene.</p><p>Outputs a sorted DataFrame containing gene names and the variances over predicted time.</p><p><strong>Required Arguments</strong></p><ul><li>cellFutures - a 3D tensor of gene expression over time; the output from predictCellFutures</li><li>geneNames - a vector of gene names corresponding to the order of the genes in cellFutures</li></ul><p><strong>Optional Arguments</strong></p><ul><li>statType - How to summarize the gene variances. Valid options are &quot;mean&quot; or &quot;median&quot;</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/rossinerbe/RNAForecaster.jl/blob/0151583ae30942235017fc2fb120fe8914a03ebb/src/makeRecursivePredictions.jl#L171-L186">source</a></section></article></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../">« Home</a><a class="docs-footer-nextpage" href="../splicedData/">Predictions from Splicing Data »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.23 on <span class="colophon-date" title="Tuesday 1 November 2022 17:13">Tuesday 1 November 2022</span>. Using Julia version 1.8.2.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
