<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Predictions from Splicing Data · RNAForecaster</title><script data-outdated-warner src="../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.045/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.24/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../">RNAForecaster</a></span></div><form class="docs-search" action="../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../">Home</a></li><li><a class="tocitem" href="../training/">Training and Forecasting</a></li><li class="is-active"><a class="tocitem" href>Predictions from Splicing Data</a></li><li><a class="tocitem" href="../forecastMLData/">Metabolic Labeling for Forecasts</a></li><li><a class="tocitem" href="../functions/">Functions</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Predictions from Splicing Data</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Predictions from Splicing Data</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/rossinerbe/RNAForecaster.jl/blob/main/docs/src/splicedData.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Predictions-from-Splicing-Data"><a class="docs-heading-anchor" href="#Predictions-from-Splicing-Data">Predictions from Splicing Data</a><a id="Predictions-from-Splicing-Data-1"></a><a class="docs-heading-anchor-permalink" href="#Predictions-from-Splicing-Data" title="Permalink"></a></h1><p>RNAForecaster requires transcriptomics data from the same cell at two different points in time. The best way to acquire this sort of information from standard single cell RNA-seq protocols is to leverage incidental capture of introns to determine whether a transcript is spliced or unspliced. In general, it is a reasonable assumption that spliced transcripts were produced more recently than unspliced, thus giving some temporal information on the sequenced transcripts.</p><p>Unfortunately, the spliced and unspliced count matrices thus generated cannot be considered two equivalent time points as the count matrices come from very different distributions, which prevents us from being able to straightforwardly predict future expression states from one time point to the next. We could attempt to handle this problem using RNA velocity style assumptions about the splicing and degradation rates to update our predicted expression levels at each time point. However, with RNAForecaster this process will propagate too much error for the resulting predictions to be of value.</p><p>Instead, we make a single prediction about the direction of regulation, which can be compared to predictions under perturbations, allowing us to uncover key genes relating to the biological process of interest.</p><p>Here we will use a small subset of murine pancreatic development data published by <a href="https://journals.biologists.com/dev/article/146/12/dev173849/19483/Comprehensive-single-cell-mRNA-profiling-reveals-a">Bastidas-Ponce et al. (2019)</a> to demonstrate how RNAForecaster is used with splicing data.</p><pre><code class="language-julia hljs">#read in data
using DelimitedFiles
splicedSub = readdlm(&quot;data/pancSplicedExampleData.csv&quot;, &#39;,&#39;)
unsplicedSub = readdlm(&quot;data/pancUnsplicedExampleData.csv&quot;, &#39;,&#39;)
geneNamesSub = readdlm(&quot;data/pancGeneNamesExampleData.csv&quot;, &#39;,&#39;)
geneNamesSub = string.(geneNamesSub)[:,1]

#data must be log normalized and converted to Float32
splicedSub = Float32.(log1p.(splicedSub))
unsplicedSub = Float32.(log1p.(unsplicedSub))

#train RNAForecaster
trainedModel = trainRNAForecaster(splicedSub, unsplicedSub);</code></pre><p>So we now have a neural network that can predicted unspliced counts based on spliced counts. We can now perturb each gene to see how the predictions of unspliced counts change. This can help uncover regulatory relationships and important genes underlying the system&#39;s biology.</p><p>Here we will simulate a KO of each gene of the 10 genes in the data from the initial conditions of 25 random cells.</p><pre><code class="language-julia hljs">KOEffects = perturbEffectPredictions(trainedModel[1], splicedSub, 25);</code></pre><p>We can search these results for statistically significant effects.</p><pre><code class="language-julia hljs">findSigRegulation(KOEffects, geneNamesSub)</code></pre><p>We can also specify specific genes and specific perturbations.</p><pre><code class="language-julia hljs">perturbEffects = perturbEffectPredictions(trainedModel[1], splicedSub, 25,
  perturbGenes = geneNamesSub[[2,5]], geneNames = geneNamesSub,
  perturbLevels = [1.0f0, 1.5f0]);</code></pre><article class="docstring"><header><a class="docstring-binding" id="RNAForecaster.perturbEffectPredictions" href="#RNAForecaster.perturbEffectPredictions"><code>RNAForecaster.perturbEffectPredictions</code></a> — <span class="docstring-category">Function</span></header><section><div><p><code>perturbEffectPredictions(trainedNetwork, splicedData::Matrix{Float32}, nCells::Int;      perturbGenes::Vector{String} = Vector{String}(undef, 0), geneNames::Vector{String} = Vector{String}(undef,0),      seed::Int=123)</code></p><p>Based on spliced/unspliced counts, predict the immediate transcriptomic effect of any or all single gene perturbations. Outputs a tuple containing the cells used for prediction, the expression predictions, the gene wise differences, and the cell-wise euclidean distances for each perturbation.</p><p>This function is capable of running on multiple parallel processes using Distributed.jl. Call addprocs(n) before running the function to add parallel workers, where n is the number of additional processes desired.</p><p><strong>Required Arguments</strong></p><ul><li>trainedNetwork - trained neuralODE from trainRNAForecaster</li><li>splicedData - log normalized spliced counts matrix. Must be in Float32 format</li><li>nCells - how many cells from the data should be used for prediction of perturb effect.</li></ul><p>Higher values will increase computational time required.</p><p><strong>Optional Arguments</strong></p><ul><li>perturbGenes - list of genes to simulate a perturbation of. By default all genes are used</li><li>geneNames - if providing a subset of the genes to perturb, a vector of gene names to</li></ul><p>match against, in the order of splicedData</p><ul><li>perturbLevels - list of perturbation levels to use for each perturbed gene. By default</li></ul><p>all genes are set to zero, simulating a KO.</p><ul><li>seed - Random seed for reproducibility on the cells chosen for prediction</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/rossinerbe/RNAForecaster.jl/blob/0151583ae30942235017fc2fb120fe8914a03ebb/src/splicedDataPerturbationEffectPredictions.jl#L1-L27">source</a></section></article><p>Convenience functions are provided to search the results of the perturbations</p><article class="docstring"><header><a class="docstring-binding" id="RNAForecaster.totalPerturbImpact" href="#RNAForecaster.totalPerturbImpact"><code>RNAForecaster.totalPerturbImpact</code></a> — <span class="docstring-category">Function</span></header><section><div><p><code>totalPerturbImpact(perturbData, geneNames::Vector{String})</code></p><p>Function to yield a sorted data frame of the size of the predicted effect of a perturbation on the cellular transcriptome. Intended to serve as a measure of more or less impactful gene perturbations.</p><p><strong>Required Arguments</strong></p><ul><li>perturbData - results from perturbEffectPredictions function</li><li>geneNames - vector of gene names in the order of the input expression data.</li></ul><p>Should only include perturbed genes</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/rossinerbe/RNAForecaster.jl/blob/0151583ae30942235017fc2fb120fe8914a03ebb/src/splicedDataPerturbationEffectPredictions.jl#L124-L135">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="RNAForecaster.genePerturbExpressionChanges" href="#RNAForecaster.genePerturbExpressionChanges"><code>RNAForecaster.genePerturbExpressionChanges</code></a> — <span class="docstring-category">Function</span></header><section><div><p><code>genePerturbExpressionChanges(perturbData, geneNames::Vector{String}, perturbGene::String;     genesperturbd::Vector{String} = geneNames)</code></p><p>Function to get a sorted data frame of the predicted effect of a gene perturb on all other genes.</p><p><strong>Required Arguments</strong></p><ul><li>perturbData - results from perturbEffectPredictions function</li><li>geneNames - vector of gene names in the order of the input expression data</li><li>perturbGene - a gene name to query the predicted perturb effect on expression</li></ul><p><strong>Optional Arguments</strong></p><ul><li>genesPerturbed - If less than all the gene perturbs were performed, the ordered names of the perturb genes must be supplied</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/rossinerbe/RNAForecaster.jl/blob/0151583ae30942235017fc2fb120fe8914a03ebb/src/splicedDataPerturbationEffectPredictions.jl#L149-L162">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="RNAForecaster.geneResponseToPerturb" href="#RNAForecaster.geneResponseToPerturb"><code>RNAForecaster.geneResponseToPerturb</code></a> — <span class="docstring-category">Function</span></header><section><div><p><code>geneResponseToPerturb(perturbData, geneNames::Vector{String}, geneOfInterest::String;     genesPerturbed::Vector{String} = geneNames)</code></p><p>Function to get a sorted data frame of the predicted effect of all other gene perturbations on a particular gene of interest.</p><p><strong>Required Arguments</strong></p><ul><li>perturbData - results from perturbEffectPredictions function</li><li>geneNames - vector of gene names in the order of the input expression data</li><li>geneOfInterest - a gene name to query</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/rossinerbe/RNAForecaster.jl/blob/0151583ae30942235017fc2fb120fe8914a03ebb/src/splicedDataPerturbationEffectPredictions.jl#L195-L205">source</a></section></article><p>For example, we can get a list of the genes, and the size of the impact of the KO on the rest of the genes&#39; expression, measured in euclidean distance, using</p><pre><code class="language-julia hljs">totalPerturbImpact(KOEffects, geneNamesSub)</code></pre></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../training/">« Training and Forecasting</a><a class="docs-footer-nextpage" href="../forecastMLData/">Metabolic Labeling for Forecasts »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.23 on <span class="colophon-date" title="Tuesday 1 November 2022 17:13">Tuesday 1 November 2022</span>. Using Julia version 1.8.2.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
